p2p/peerResend.go

```diff
@@ -0,0 +1,99 @@
+package p2p
+
+import (
+	"crypto/sha1"
+	"sync"
+	"time"
+)
+
+type PeerResend struct {
+	mtx      sync.RWMutex
+	buckets  []*PRBucket
+	pc       int
+	interval time.Duration
+	stopper  chan interface{}
+}
+
+type PRBucket struct {
+	mtx  sync.RWMutex
+	data map[[sha1.Size]byte]bool
+}
+
+func (b *PRBucket) Has(hash [sha1.Size]byte) bool {
+	b.mtx.RLock()
+	defer b.mtx.RUnlock()
+	return b.data[hash]
+}
+
+func (b *PRBucket) Add(hash [sha1.Size]byte) {
+	b.mtx.Lock()
+	defer b.mtx.Unlock()
+	b.data[hash] = true
+}
+
+func NewPeerResend(buckets int, interval time.Duration) *PeerResend {
+	pr := new(PeerResend)
+	pr.interval = interval
+	pr.buckets = make([]*PRBucket, buckets)
+	for i := 0; i < buckets; i++ {
+		pr.buckets[i] = newPRBucket()
+	}
+	pr.stopper = make(chan interface{}, 1)
+	go pr.Cleanup()
+	return pr
+}
+
+func newPRBucket() *PRBucket {
+	b := new(PRBucket)
+	b.data = make(map[[sha1.Size]byte]bool)
+	return b
+}
+
+func (pr *PeerResend) Has(hash [sha1.Size]byte) bool {
+	pr.mtx.RLock()
+	defer pr.mtx.RUnlock()
+
+	for i := 0; i < len(pr.buckets); i++ {
+		if pr.buckets[pr.ptr(i)].Has(hash) {
+			return true
+		}
+	}
+	return false
+}
+
+func (pr *PeerResend) Add(hash [sha1.Size]byte) {
+	pr.mtx.RLock()
+	defer pr.mtx.RUnlock()
+	pr.buckets[pr.pc].Add(hash)
+}
+
+func (pr *PeerResend) Stop() {
+	close(pr.stopper)
```

Comment:
should we protect this with a Do.once? Because calling close on an already closed channel cause a panic I think

Reply:
Hm that's a tough call. It is idiomatic Golang to have things panic if closed twice, leaving it up to the implementer to make sure they don't close it twice in exchange for lower overhead.

Reply:
Ok, I see you added a comment on the method, that makes sense